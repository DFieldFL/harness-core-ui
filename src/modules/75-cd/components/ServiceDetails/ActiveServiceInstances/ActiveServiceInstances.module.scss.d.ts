/* eslint-disable */
// this is an auto-generated file
declare const styles: {
  readonly activeServiceInstances: string
  readonly activeServiceInstancesHeader: string
  readonly instanceCount: string
  readonly pieChartLabelContainerStyles: string
  readonly pieChartLabelStyles: string
  readonly instanceTable: string
  readonly paddedContainer: string
  readonly environmentRow: string
  readonly overflow: string
  readonly seeMore: string
  readonly hexContainer: string
  readonly hex: string
  readonly hover: string
  readonly activeServiceInstancesEmpty: string
  readonly activeServiceInstancePopover: string
  readonly activeServiceInstancePopoverSection: string
  readonly sectionValue: string
}
export default styles
