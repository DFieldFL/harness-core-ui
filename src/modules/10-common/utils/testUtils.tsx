import React from 'react'
import { UseGetProps, UseGetReturn, RestfulProvider } from 'restful-react'
import { compile } from 'path-to-regexp'
import { createMemoryHistory } from 'history'
import { Router, Route, Switch, useLocation } from 'react-router-dom'
import { ModalProvider } from '@wings-software/uikit'
import qs from 'qs'

import strings from 'strings/strings.en.yaml'
import { AppStoreProvider, useAppStoreWriter } from 'framework/hooks/useAppStore'
import type { AppStore } from 'framework/types/AppStore'
import { AUTH_ROUTE_PATH_PREFIX, StringsContext } from 'framework/exports'

export type UseGetMockData<TData, TError = undefined, TQueryParams = undefined, TPathParams = undefined> = Required<
  UseGetProps<TData, TError, TQueryParams, TPathParams>
>['mock']

export interface UseMutateMockData<TData, TRequestBody = unknown> {
  loading?: boolean
  mutate?: (data?: TRequestBody) => Promise<TData>
}

export type UseGetReturnData<TData, TError = undefined, TQueryParams = undefined, TPathParams = undefined> = Omit<
  UseGetReturn<TData, TError, TQueryParams, TPathParams>,
  'absolutePath' | 'cancel' | 'response'
>

export const findDialogContainer = (): Element | null => document.querySelector('.bp3-dialog')
export const findPopoverContainer = (): Element | null => document.querySelector('.bp3-popover-content')

interface TestWrapperProps {
  path?: string
  pathParams?: Record<string, string | number>
  queryParams?: Record<string, any>
  defaultAppStoreValues?: Partial<AppStore>
}

export const prependAccountPath = (path: string): string => AUTH_ROUTE_PATH_PREFIX + path

const AppStoreInitializer: React.FC<{ defaultAppStoreValues: TestWrapperProps['defaultAppStoreValues'] }> = ({
  children,
  defaultAppStoreValues
}) => {
  const updateAppStore = useAppStoreWriter()
  updateAppStore(defaultAppStoreValues || {})
  return <> {children} </>
}

export const NotFound = (): JSX.Element => {
  const location = useLocation()
  return (
    <div>
      <h1>Not Found</h1>
      <div data-testid="location">{`${location.pathname}${
        location.search ? `?${location.search.replace(/^\?/g, '')}` : ''
      }`}</div>
    </div>
  )
}

export const TestWrapper: React.FC<TestWrapperProps> = props => {
  const { path = '/', pathParams = {}, defaultAppStoreValues, queryParams = {} } = props

  const search = qs.stringify(queryParams, { addQueryPrefix: true })
  const routePath = compile(path)(pathParams) + search

  const history = createMemoryHistory({ initialEntries: [routePath] })

  return (
    <AppStoreProvider>
      <StringsContext.Provider value={strings}>
        <AppStoreInitializer defaultAppStoreValues={defaultAppStoreValues}>
          <Router history={history}>
            <ModalProvider>
              <RestfulProvider base="/">
                <Switch>
                  <Route exact path={path}>
                    {props.children}
                  </Route>
                  <Route>
                    <NotFound />
                  </Route>
                </Switch>
              </RestfulProvider>
            </ModalProvider>
          </Router>
        </AppStoreInitializer>
      </StringsContext.Provider>
    </AppStoreProvider>
  )
}
