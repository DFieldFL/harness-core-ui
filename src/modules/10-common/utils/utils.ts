import type { IconName } from '@wings-software/uicore'
import { Scope } from '@common/interfaces/SecretsInterface'
import { ModuleName } from 'framework/types/ModuleName'

interface SetPageNumberProps {
  setPage: (value: React.SetStateAction<number>) => void
  pageItemsCount?: number
  page: number
}

export const getModuleIcon = (module: ModuleName): IconName => {
  switch (module) {
    case ModuleName.CD:
      return 'cd-main'
    case ModuleName.CV:
      return 'cv-main'
    case ModuleName.CI:
      return 'ci-main'
    case ModuleName.CE:
      return 'ce-main'
    case ModuleName.CF:
      return 'cf-main'
  }
  return 'nav-project'
}

export const getReference = (scope?: Scope, identifier?: string): string | undefined => {
  switch (scope) {
    case Scope.PROJECT:
      return identifier
    case Scope.ORG:
      return `org.${identifier}`
    case Scope.ACCOUNT:
      return `account.${identifier}`
  }
}

export const setPageNumber = ({ setPage, pageItemsCount, page }: SetPageNumberProps): void => {
  if (pageItemsCount === 0 && page > 0) {
    setPage(page - 1)
  }
}
