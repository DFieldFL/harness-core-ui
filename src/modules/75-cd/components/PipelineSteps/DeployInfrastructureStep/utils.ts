/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { FormikErrors, yupToFormErrors } from 'formik'
import { defaultTo, get, isEmpty } from 'lodash-es'
import * as Yup from 'yup'

import {
  getMultiTypeFromValue,
  MultiTypeInputType,
  RUNTIME_INPUT_VALUE,
  SelectOption,
  MultiSelectWithSubmenuOption
} from '@harness/uicore'
import type { EnvironmentResponseDTO } from 'services/cd-ng'

import { getDurationValidationSchema } from '@common/components/MultiTypeDuration/MultiTypeDuration'

import type { DeployStageConfig } from '@pipeline/utils/DeployStageInterface'
import { StepViewType, ValidateInputSetProps } from '@pipeline/components/AbstractSteps/Step'

export interface DeployInfrastructureProps {
  initialValues: DeployStageConfig
  onUpdate?: (data: DeployStageConfig) => void
  readonly: boolean
  allowableTypes: MultiTypeInputType[]
  stepViewType?: StepViewType
  inputSetData?: {
    template?: DeployStageConfig
    path?: string
    readonly?: boolean
  }
}

export const ALL_SELECTED = 'All'

export function isEditEnvironment(data?: EnvironmentResponseDTO): boolean {
  if (!isEmpty(data?.identifier)) {
    return true
  }
  return false
}

export function isEditEnvironmentOrEnvGroup(data?: EnvironmentResponseDTO): boolean {
  if (!isEmpty(data?.identifier)) {
    return true
  }
  return false
}

export function isEditInfrastructure(data?: string): boolean {
  if (!isEmpty(data)) {
    return true
  }
  return false
}

export function validateStepForm({
  data,
  template,
  getString,
  viewType
}: ValidateInputSetProps<any>): FormikErrors<any> {
  const isRequired = viewType === StepViewType.DeploymentForm || viewType === StepViewType.TriggerForm
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const errors = {} as any
  // istanbul ignore next
  // istanbul ignore else
  if (getMultiTypeFromValue(template?.timeout) === MultiTypeInputType.RUNTIME) {
    // istanbul ignore next
    let timeoutSchema = getDurationValidationSchema({ minimum: '10s' })
    // istanbul ignore next
    if (isRequired) {
      // istanbul ignore next
      timeoutSchema = timeoutSchema.required(getString?.('validation.timeout10SecMinimum'))
    }
    const timeout = Yup.object().shape({
      timeout: timeoutSchema
    })

    try {
      timeout.validateSync(data)
    } catch (e) {
      /* istanbul ignore else */
      if (e instanceof Yup.ValidationError) {
        const err = yupToFormErrors(e)

        Object.assign(errors, err)
      }
    }
  }
  return errors
}

export function processNonGitOpsInitialValues(initialValues: DeployStageConfig) {
  return {
    environment: {
      environmentRef: defaultTo(initialValues.environment?.environmentRef, ''),
      deployToAll: defaultTo(initialValues.environment?.deployToAll, false)
    },
    infrastructureRef: (initialValues.environment?.infrastructureDefinitions?.[0].ref ||
      initialValues.environment?.infrastructureDefinitions ||
      '') as string
  }
}

export function processNonGitOpsFormValues(data: DeployStageConfig) {
  return {
    environment: {
      environmentRef: data.environment?.environmentRef,
      deployToAll: data.environment?.environmentRef === RUNTIME_INPUT_VALUE ? true : false,
      ...(data.environment?.environmentRef &&
        data.environment?.environmentRef !== RUNTIME_INPUT_VALUE &&
        data.infrastructureRef && {
          infrastructureDefinitions:
            data.infrastructureRef === RUNTIME_INPUT_VALUE
              ? RUNTIME_INPUT_VALUE
              : [
                  {
                    ref: data.infrastructureRef
                  }
                ]
        })
    }
  }
}

export function processGitOpsEnvironmentInitialValues(initialValues: DeployStageConfig) {
  return {
    isEnvGroup: false,
    environmentOrEnvGroupRef: defaultTo(initialValues.environment?.environmentRef, ''),
    environmentOrEnvGroupAsRuntime: 'Environment',
    deployToAll: defaultTo(initialValues.environment?.deployToAll, false),
    clusterRef:
      getMultiTypeFromValue(initialValues.environment?.gitOpsClusters as any) === MultiTypeInputType.RUNTIME
        ? RUNTIME_INPUT_VALUE
        : initialValues.environment?.deployToAll
        ? [
            {
              label: 'All Clusters in Environment',
              value: ALL_SELECTED
            }
          ]
        : defaultTo(initialValues.environment?.gitOpsClusters, [])?.map(cluster => {
            return {
              label: cluster.ref,
              value: cluster.ref
            }
          })
  }
}

export function processGitOpsEnvironmentFormValues(data: DeployStageConfig) {
  const allClustersSelected = (data.clusterRef as SelectOption[])?.[0].value === ALL_SELECTED

  return {
    environment: {
      environmentRef:
        data.environmentOrEnvGroupRef === RUNTIME_INPUT_VALUE
          ? RUNTIME_INPUT_VALUE
          : defaultTo((data.environmentOrEnvGroupRef as SelectOption)?.value, ''),
      deployToAll: data.environmentOrEnvGroupRef === RUNTIME_INPUT_VALUE ? true : allClustersSelected,
      ...(data.environmentOrEnvGroupRef &&
        data.environmentOrEnvGroupRef !== RUNTIME_INPUT_VALUE &&
        !allClustersSelected && {
          gitOpsClusters:
            data.clusterRef === RUNTIME_INPUT_VALUE
              ? RUNTIME_INPUT_VALUE
              : (data.clusterRef as SelectOption[])?.map(cluster => ({ ref: cluster.value }))
        })
    }
  }
}

export function processGitOpsEnvGroupInitialValues(initialValues: DeployStageConfig) {
  return {
    isEnvGroup: true,
    environmentOrEnvGroupRef: defaultTo(initialValues.environmentGroup?.envGroupRef, ''),
    environmentOrEnvGroupAsRuntime: 'Environment Group',
    environmentsInEnvGroup:
      getMultiTypeFromValue(initialValues.environmentGroup?.envGroupRef) === MultiTypeInputType.RUNTIME
        ? RUNTIME_INPUT_VALUE
        : initialValues.environmentGroup?.deployToAll
        ? [
            {
              name: ALL_SELECTED
            }
          ]
        : defaultTo(initialValues.environmentGroup?.envGroupConfig, [])?.map(environment => {
            return {
              name: environment.environmentRef,
              deployToAll: environment.deployToAll,
              ...(!environment.deployToAll && { clusters: defaultTo(environment.gitOpsClusters, []) })
            }
          }),
    deployToAll: defaultTo(initialValues.environmentGroup?.deployToAll, false),
    clusterRef: initialValues.environmentGroup?.deployToAll
      ? RUNTIME_INPUT_VALUE
      : initialValues.environmentGroup?.envGroupConfig?.reduce((prev, environment) => {
          return [
            ...prev,
            ...(environment.deployToAll
              ? [
                  {
                    label: `All`,
                    value: 'All',
                    parentLabel: environment.environmentRef,
                    parentValue: environment.environmentRef
                  }
                ]
              : defaultTo(
                  environment?.gitOpsClusters?.map(cluster => ({
                    label: cluster.ref,
                    value: cluster.ref,
                    parentLabel: environment.environmentRef,
                    parentValue: environment.environmentRef
                  })),
                  []
                ))
          ]
        }, [] as MultiSelectWithSubmenuOption[])
  }
}

export function processGitOpsEnvGroupFormValues(data: DeployStageConfig) {
  const environmentClusterMap: any = {}
  ;(data.clusterRef as MultiSelectWithSubmenuOption[])?.forEach(cluster => {
    if (cluster.value !== 'All') {
      try {
        environmentClusterMap[get(cluster, 'parentValue', '')].push(cluster.value)
      } catch (e: any) {
        environmentClusterMap[get(cluster, 'parentValue', '')] = [cluster.value]
      }
    }
  })

  const allEnvironmentsSelected = (data.environmentInEnvGroupRef as SelectOption[])?.[0].value === ALL_SELECTED

  return {
    environmentGroup: {
      envGroupRef:
        data.environmentOrEnvGroupRef === RUNTIME_INPUT_VALUE
          ? RUNTIME_INPUT_VALUE
          : defaultTo((data.environmentOrEnvGroupRef as SelectOption)?.value, ''),
      deployToAll: data.environmentOrEnvGroupRef === RUNTIME_INPUT_VALUE || allEnvironmentsSelected,
      ...(!(data.environmentOrEnvGroupRef === RUNTIME_INPUT_VALUE || allEnvironmentsSelected) && {
        envGroupConfig:
          data.environmentInEnvGroupRef === RUNTIME_INPUT_VALUE
            ? RUNTIME_INPUT_VALUE
            : (data?.environmentInEnvGroupRef as SelectOption[])?.map(environmentInEnvGroup => {
                const areClustersSelected = Boolean(environmentClusterMap[environmentInEnvGroup.value])
                return {
                  environmentRef: environmentInEnvGroup?.value,
                  deployToAll: !areClustersSelected,
                  ...(areClustersSelected && {
                    gitOpsClusters:
                      data.clusterRef === RUNTIME_INPUT_VALUE
                        ? RUNTIME_INPUT_VALUE
                        : environmentClusterMap[environmentInEnvGroup.value]?.map((cluster: any) => ({ ref: cluster }))
                  })
                }
              })
      })
    }
  }
}
