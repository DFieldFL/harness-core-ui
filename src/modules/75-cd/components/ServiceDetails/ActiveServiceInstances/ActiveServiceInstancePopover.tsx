import React from 'react'
import { useParams } from 'react-router-dom'
import { Card, Color, Layout, Text } from '@wings-software/uicore'
import {
  GetActiveInstancesByServiceIdEnvIdAndBuildIdsQueryParams,
  InstanceDetailsDTO,
  useGetActiveInstancesByServiceIdEnvIdAndBuildIds
} from 'services/cd-ng'
import type { ProjectPathProps, ServicePathProps } from '@common/interfaces/RouteInterfaces'
import { useStrings } from 'framework/strings'
import css from '@cd/components/ServiceDetails/ActiveServiceInstances/ActiveServiceInstances.module.scss'

export type InstanceData = Record<string, Record<string, InstanceDetailsDTO[]>>

export interface ActiveServiceInstancePopoverProps {
  buildId?: string
  envId?: string
  instanceNum?: number
}

interface SectionProps {
  header: string
  values: {
    label: string
    value: string
  }[]
}

const Section: React.FC<{ data: SectionProps[] }> = props => {
  const { data } = props
  return (
    <Card className={css.activeServiceInstancePopover}>
      <Layout.Vertical>
        {data.map(item => {
          return (
            <Layout.Vertical key={item.header} className={css.activeServiceInstancePopoverSection}>
              <Text font={{ weight: 'semi-bold', size: 'small' }} margin={{ bottom: 'small' }} color={Color.GREY_800}>
                {item.header}
              </Text>
              <Layout.Vertical>
                {item.values.map(itemValue =>
                  itemValue.value ? (
                    <Layout.Horizontal key={itemValue.label}>
                      <Text
                        font={{ weight: 'semi-bold', size: 'small' }}
                        color={Color.GREY_500}
                        margin={{ right: 'medium', bottom: 'xsmall' }}
                      >
                        {`${itemValue.label}:`}
                      </Text>
                      <Text
                        font={{ weight: 'semi-bold', size: 'small' }}
                        color={Color.GREY_800}
                        className={css.sectionValue}
                      >
                        {itemValue.value}
                      </Text>
                    </Layout.Horizontal>
                  ) : (
                    <></>
                  )
                )}
              </Layout.Vertical>
            </Layout.Vertical>
          )
        })}
      </Layout.Vertical>
    </Card>
  )
}

export const ActiveServiceInstancePopover: React.FC<ActiveServiceInstancePopoverProps> = props => {
  const { buildId = '', envId = '', instanceNum = 0 } = props
  const { accountId, orgIdentifier, projectIdentifier, serviceId } = useParams<ProjectPathProps & ServicePathProps>()
  const { getString } = useStrings()

  const queryParams: GetActiveInstancesByServiceIdEnvIdAndBuildIdsQueryParams = {
    accountIdentifier: accountId,
    orgIdentifier,
    projectIdentifier,
    serviceId,
    envId,
    buildIds: [buildId]
  }
  const { loading, data, error } = useGetActiveInstancesByServiceIdEnvIdAndBuildIds({
    queryParams,
    queryParamStringifyOptions: {
      arrayFormat: 'repeat'
    }
  })

  if (loading || error || !data?.data?.instancesByBuildIdList?.[0]?.instances?.length) {
    return <></>
  }

  const instanceData = data?.data?.instancesByBuildIdList?.[0]?.instances[instanceNum] || {}

  const sectionData: SectionProps[] = [
    {
      header: getString('cd.serviceDashboard.instanceDetails'),
      values: [
        {
          label: getString('cd.serviceDashboard.pod'),
          value: instanceData.podName || ''
        },
        {
          label: getString('cd.serviceDashboard.artifact'),
          value: instanceData.artifactName || ''
        }
      ]
    },
    {
      header: getString('infrastructureText'),
      values: Object.keys(instanceData.infrastructureDetails || {}).map(infrastructureDetailsKey => ({
        label: infrastructureDetailsKey,
        value: instanceData.infrastructureDetails?.[infrastructureDetailsKey]
      }))
    },
    {
      header: getString('deploymentText'),
      values: [
        {
          label: getString('cd.serviceDashboard.deployedAt'),
          value: instanceData.deployedAt
        },
        {
          label: getString('cd.serviceDashboard.deployedBy'),
          value: instanceData.deployedByName
        },
        {
          label: getString('common.pipeline'),
          value: instanceData.pipelineExecutionName
        }
      ]
    }
  ]

  return <Section data={sectionData}></Section>
}
