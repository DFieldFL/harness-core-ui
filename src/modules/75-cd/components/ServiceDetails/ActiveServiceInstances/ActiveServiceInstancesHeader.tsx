import React, { useMemo } from 'react'
import { useParams } from 'react-router-dom'
import moment from 'moment'
import { Color, Container, Layout, Text } from '@wings-software/uicore'
import type { ProjectPathProps, ServicePathProps } from '@common/interfaces/RouteInterfaces'
import {
  GetActiveServiceInstanceSummaryQueryParams,
  GetInstanceGrowthTrendQueryParams,
  useGetActiveServiceInstanceSummary,
  useGetInstanceGrowthTrend
} from 'services/cd-ng'
import { PieChart, PieChartProps } from '@cd/components/PieChart/PieChart'
import { useStrings } from 'framework/strings'
import { numberFormatter } from '@cd/components/Services/common'
import { TrendPopover } from '@cd/components/TrendPopover/TrendPopover'
import { SparklineChart } from '@common/components/SparklineChart/SparklineChart'
import { Ticker } from '@common/components/Ticker/Ticker'
import css from '@cd/components/ServiceDetails/ActiveServiceInstances/ActiveServiceInstances.module.scss'

export const ActiveServiceInstancesHeader: React.FC = () => {
  const { getString } = useStrings()
  const { accountId, orgIdentifier, projectIdentifier, serviceId } = useParams<ProjectPathProps & ServicePathProps>()

  const queryParams: GetActiveServiceInstanceSummaryQueryParams = useMemo(
    () => ({
      accountIdentifier: accountId,
      orgIdentifier,
      projectIdentifier,
      serviceId,
      timestamp: moment().subtract(30, 'days').toDate().getTime()
    }),
    [accountId, orgIdentifier, projectIdentifier, serviceId]
  )
  const { data, error } = useGetActiveServiceInstanceSummary({ queryParams })

  const instanceGrowthTrendQueryParams: GetInstanceGrowthTrendQueryParams = useMemo(
    () => ({
      accountIdentifier: accountId,
      orgIdentifier,
      projectIdentifier,
      serviceId,
      startTime: moment().utc().startOf('day').subtract(6, 'months').toDate().getTime(),
      endTime: moment().utc().startOf('day').toDate().getTime()
    }),
    [accountId, orgIdentifier, projectIdentifier, serviceId]
  )
  const { data: instanceGrowthTrendData } = useGetInstanceGrowthTrend({ queryParams: instanceGrowthTrendQueryParams })

  const trendData: number[] = useMemo(() => {
    const timeValuePairList = instanceGrowthTrendData?.data?.timeValuePairList || []
    if (!timeValuePairList.length) {
      return []
    }
    timeValuePairList.sort((prev, curr) => (prev.timestamp || 0) - (curr.timestamp || 0))
    return timeValuePairList.map(timeValuePair => timeValuePair.value || 0)
  }, [instanceGrowthTrendData])

  if (error) {
    return <></>
  }

  const { nonProdInstances = 0, prodInstances = 0, totalInstances = 0 } = data?.data?.countDetails || {}

  const changeRate = data?.data?.changeRate || 0

  const pieChartProps: PieChartProps = {
    items: [
      {
        label: getString('cd.serviceDashboard.nonProd'),
        value: nonProdInstances,
        formattedValue: numberFormatter(nonProdInstances),
        color: 'var(--primary-2)'
      },
      {
        label: getString('cd.serviceDashboard.prod'),
        value: prodInstances,
        formattedValue: numberFormatter(prodInstances),
        color: 'var(--primary-7)'
      }
    ],
    size: 36,
    labelContainerStyles: css.pieChartLabelContainerStyles,
    labelStyles: css.pieChartLabelStyles,
    options: {
      tooltip: {
        enabled: false
      }
    }
  }

  const tickerColor = changeRate > 0 ? Color.GREEN_600 : Color.RED_500

  return (
    <Layout.Horizontal
      flex={{ alignItems: 'center', justifyContent: 'space-between' }}
      padding={{ bottom: 'small' }}
      className={css.activeServiceInstancesHeader}
    >
      <Layout.Horizontal>
        <Text
          font={{ weight: 'bold' }}
          color={Color.GREY_800}
          className={css.instanceCount}
          margin={{ right: 'large' }}
        >
          {numberFormatter(totalInstances)}
        </Text>
        <Layout.Vertical flex={{ alignItems: 'flex-start', justifyContent: 'flex-start' }}>
          <Ticker
            value={
              <Text font={{ size: 'xsmall' }} color={tickerColor}>{`${numberFormatter(Math.abs(changeRate), {
                truncate: false
              })}%`}</Text>
            }
            decreaseMode={changeRate < 0}
            color={tickerColor}
          />
          <Text font={{ size: 'xsmall' }}>
            {getString('cd.serviceDashboard.in', {
              timeRange: getString('cd.serviceDashboard.month').toLocaleLowerCase()
            })}
          </Text>
        </Layout.Vertical>
      </Layout.Horizontal>
      <Layout.Horizontal>
        {trendData.length ? (
          <Container margin={{ right: 'medium' }}>
            <TrendPopover data={trendData}>
              <SparklineChart
                title={getString('cd.serviceDashboard.6monthTrend')}
                data={trendData}
                options={{ chart: { width: 80, height: 50 } }}
                sparklineChartContainerStyles={css.hover}
              />
            </TrendPopover>
          </Container>
        ) : (
          <></>
        )}
        {totalInstances ? <PieChart {...pieChartProps} /> : <></>}
      </Layout.Horizontal>
    </Layout.Horizontal>
  )
}
