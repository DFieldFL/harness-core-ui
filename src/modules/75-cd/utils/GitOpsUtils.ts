import type { GitOpsInfoDTO, ManagedArgoGitOpsInfoDTO, ConnectedArgoGitOpsInfoDTO } from 'services/cd-ng'

export type GitOpsProviderType = Required<GitOpsInfoDTO>['type']

export const GitOpsProviderTypeEnum: Record<GitOpsProviderType, GitOpsProviderType> = {
  CONNECTED_ARGO_PROVIDER: 'CONNECTED_ARGO_PROVIDER',
  MANAGED_ARGO_PROVIDER: 'MANAGED_ARGO_PROVIDER'
}

export function isConnectedGitOpsProvider(obj?: GitOpsInfoDTO | null): obj is ConnectedArgoGitOpsInfoDTO {
  return obj?.type === GitOpsProviderTypeEnum.CONNECTED_ARGO_PROVIDER
}

export function isManagedGitOpsProvider(obj?: GitOpsInfoDTO | null): obj is ManagedArgoGitOpsInfoDTO {
  return obj?.type === GitOpsProviderTypeEnum.MANAGED_ARGO_PROVIDER
}
