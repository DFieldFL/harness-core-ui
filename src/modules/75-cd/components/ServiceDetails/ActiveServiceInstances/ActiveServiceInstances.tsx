import React from 'react'
import { Card, Color, Layout, Text } from '@wings-software/uicore'
import { useStrings } from 'framework/strings'
import { ActiveServiceInstancesHeader } from '@cd/components/ServiceDetails/ActiveServiceInstances/ActiveServiceInstancesHeader'
import { ActiveServiceInstancesContent } from '@cd/components/ServiceDetails/ActiveServiceInstances/ActiveServiceInstancesContent'
import css from '@cd/components/ServiceDetails/ActiveServiceInstances/ActiveServiceInstances.module.scss'

export const ActiveServiceInstances: React.FC = () => {
  const { getString } = useStrings()
  return (
    <Card className={css.activeServiceInstances}>
      <Layout.Vertical height={'100%'}>
        <Text font={{ weight: 'bold' }} color={Color.GREY_600} margin={{ bottom: 'medium' }}>
          {getString('cd.serviceDashboard.activeServiceInstancesLabel')}
        </Text>
        <ActiveServiceInstancesHeader />
        <ActiveServiceInstancesContent />
      </Layout.Vertical>
    </Card>
  )
}
