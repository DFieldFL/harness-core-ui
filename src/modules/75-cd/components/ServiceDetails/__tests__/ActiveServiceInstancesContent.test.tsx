import React from 'react'
import { render } from '@testing-library/react'
import { TestWrapper } from '@common/utils/testUtils'
import { ActiveServiceInstancesContent } from '@cd/components/ServiceDetails/ActiveServiceInstances/ActiveServiceInstancesContent'
import * as cdngServices from 'services/cd-ng'
import { envBuildInstanceCount } from '@cd/mock'

jest.mock('highcharts-react-official', () => () => <></>)

jest.spyOn(cdngServices, 'useGetEnvBuildInstanceCount').mockImplementation(() => {
  return { loading: false, error: false, data: envBuildInstanceCount, refetch: jest.fn() } as any
})

describe('ActiveServiceInstancesContent', () => {
  test('should render ActiveServiceInstancesContent', () => {
    const { container } = render(
      <TestWrapper
        path="account/:accountId/cd/orgs/:orgIdentifier/projects/:projectIdentifier/services"
        pathParams={{ accountId: 'dummy', orgIdentifier: 'dummy', projectIdentifier: 'dummy' }}
      >
        <ActiveServiceInstancesContent />
      </TestWrapper>
    )
    expect(container).toMatchSnapshot()
  })
})
