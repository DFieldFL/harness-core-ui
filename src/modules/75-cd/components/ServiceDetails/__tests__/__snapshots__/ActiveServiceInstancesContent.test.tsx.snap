// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActiveServiceInstancesContent should render ActiveServiceInstancesContent 1`] = `
<div>
  <div
    class="Layout--horizontal StyledProps--main StyledProps--padding-top-medium"
  >
    <div
      class="table instanceTable"
    >
      <div
        class="header"
        role="row"
      >
        <div
          class="cell"
          colspan="1"
          role="columnheader"
          style="width: 20%;"
        >
          <p
            class="StyledProps--font StyledProps--main StyledProps--font-variation-table-headers"
          >
            environment
          </p>
        </div>
        <div
          class="cell"
          colspan="1"
          role="columnheader"
          style="width: 20%;"
        >
          <p
            class="StyledProps--font StyledProps--main StyledProps--font-variation-table-headers"
          >
            cd.serviceDashboard.buildName
          </p>
        </div>
        <div
          class="cell"
          colspan="1"
          role="columnheader"
          style="width: 10%;"
        >
          <p
            class="StyledProps--font StyledProps--main StyledProps--font-variation-table-headers"
          >
            
          </p>
        </div>
        <div
          class="cell"
          colspan="1"
          role="columnheader"
          style="width: 50%;"
        >
          <p
            class="StyledProps--font StyledProps--main StyledProps--font-variation-table-headers"
          >
            common.instanceLabel
          </p>
        </div>
      </div>
      <div
        class="body"
      >
        <div
          class="row card"
          role="row"
        >
          <div
            class="cell"
            role="cell"
            style="width: 20%;"
          >
            <div
              class="StyledProps--font StyledProps--main paddedContainer"
            >
              <p
                class="StyledProps--font StyledProps--main environmentRow overflow StyledProps--font-size-small StyledProps--font-weight-bold StyledProps--color StyledProps--color-white"
              >
                envName
              </p>
            </div>
          </div>
          <div
            class="cell"
            role="cell"
            style="width: 20%;"
          >
            <div
              class="StyledProps--font StyledProps--main paddedContainer"
            >
              <p
                class="StyledProps--font StyledProps--main overflow StyledProps--font-size-small StyledProps--font-weight-semi-bold StyledProps--color StyledProps--color-grey800"
              >
                build1
              </p>
            </div>
          </div>
          <div
            class="cell"
            role="cell"
            style="width: 10%;"
          >
            <div
              class="StyledProps--font StyledProps--main paddedContainer"
            >
              <p
                class="StyledProps--font StyledProps--main instanceCount overflow StyledProps--font-size-xsmall StyledProps--font-weight-bold StyledProps--background-grey100"
              >
                1
              </p>
            </div>
          </div>
          <div
            class="cell"
            role="cell"
            style="width: 50%;"
          >
            <div
              class="StyledProps--font StyledProps--main paddedContainer hexContainer StyledProps--flex StyledProps--flex-justifyContent-flex-start"
            >
              <span
                class="bp3-popover-wrapper"
              >
                <span
                  class="bp3-popover-target"
                >
                  <div
                    class="StyledProps--font StyledProps--main hex StyledProps--background-primary3 StyledProps--margin-left-xsmall StyledProps--margin-right-xsmall StyledProps--margin-top-xsmall StyledProps--margin-bottom-xsmall"
                    style="width: 18px; height: 18px;"
                  />
                </span>
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
